File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockFileStaging.kt
package stasis.test.client_android.lib.mocks

import stasis.client_android.lib.staging.FileStaging
import stasis.test.client_android.lib.ResourceHelpers
import java.nio.file.FileSystem
import java.nio.file.Path
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

open class MockFileStaging : FileStaging {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.TemporaryCreated to AtomicInteger(0),
        Statistic.TemporaryDiscarded to AtomicInteger(0),
        Statistic.Destaged to AtomicInteger(0)
    )

    private var fs: FileSystem

    init {
        val (mockFs, _) = ResourceHelpers.createMockFileSystem(ResourceHelpers.FileSystemSetup.Unix)
        fs = mockFs
    }

    override suspend fun temporary(): Path {
        stats[Statistic.TemporaryCreated]?.getAndIncrement()
        return fs.getPath("/${UUID.randomUUID()}")
    }

    override suspend fun discard(file: Path) {
        stats[Statistic.TemporaryDiscarded]?.getAndIncrement()
    }

    override suspend fun destage(from: Path, to: Path) {
        stats[Statistic.Destaged]?.getAndIncrement()
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object TemporaryCreated : Statistic()
        object TemporaryDiscarded : Statistic()
        object Destaged : Statistic()
    }
}

File: client/src/test/scala/stasis/test/specs/unit/client/mocks/MockFileStaging.scala
package stasis.test.specs.unit.client.mocks

import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

import org.apache.pekko.Done
import stasis.client.staging.FileStaging
import stasis.test.specs.unit.client.ResourceHelpers
import stasis.test.specs.unit.client.mocks.MockFileStaging.Statistic

import scala.concurrent.Future

class MockFileStaging() extends FileStaging with ResourceHelpers {
  private val stats: Map[Statistic, AtomicInteger] = Map(
    Statistic.TemporaryCreated -> new AtomicInteger(0),
    Statistic.TemporaryDiscarded -> new AtomicInteger(0),
    Statistic.Destaged -> new AtomicInteger(0)
  )

  val temporaryPath: Path = "/ops/temp-file-1".asTestResource

  override def temporary(): Future[Path] = {
    stats(Statistic.TemporaryCreated).incrementAndGet()
    Future.successful(temporaryPath)
  }

  override def discard(file: Path): Future[Done] = {
    stats(Statistic.TemporaryDiscarded).incrementAndGet()
    Future.successful(Done)
  }

  override def destage(from: Path, to: Path): Future[Done] = {
    stats(Statistic.Destaged).incrementAndGet()
    Future.successful(Done)
  }

  def statistics: Map[Statistic, Int] = stats.view.mapValues(_.get()).toMap
}

object MockFileStaging {
  sealed trait Statistic
  object Statistic {
    case object TemporaryCreated extends Statistic
    case object TemporaryDiscarded extends Statistic
    case object Destaged extends Statistic
  }
}
